### PR TITLE
In .run(fn), check if fn exists before executing the callback, fixes #1496

### LINKED ISSUE #1496: Mocha.run throws a TypeError if callback is not provided
To reproduce the issue, install Mocha 2.1.0 and run this script on Node console.
```javascript
var Mocha = require("mocha");
var mocha = new Mocha();
mocha.addFile("./test.js");
mocha.run();
```

Here is `test.js`
```javascript
it("nothing", function() {});
```

Then, I get:
```
Donghwan@DONGHWAN ~/Documents/GitHub/vibe-javascript-client (master)
$ node test-runner


  √ nothing

  1 passing (7ms)


c:\Users\Donghwan\Documents\GitHub\vibe-javascript-client\node_modules\mocha\lib\mocha.js:414
          fn(failures);
          ^
TypeError: undefined is not a function
    at done (c:\Users\Donghwan\Documents\GitHub\vibe-javascript-client\node_modules\mocha\lib\mocha.js:414:11)
    at Runner.<anonymous> (c:\Users\Donghwan\Documents\GitHub\vibe-javascript-client\node_modules\mocha\lib\runner.js:590:5)
    at Runner.emit (events.js:117:20)
    at c:\Users\Donghwan\Documents\GitHub\vibe-javascript-client\node_modules\mocha\lib\runner.js:597:10
    at c:\Users\Donghwan\Documents\GitHub\vibe-javascript-client\node_modules\mocha\lib\runner.js:518:7
    at next (c:\Users\Donghwan\Documents\GitHub\vibe-javascript-client\node_modules\mocha\lib\runner.js:248:23)
    at Object._onImmediate (c:\Users\Donghwan\Documents\GitHub\vibe-javascript-client\node_modules\mocha\lib\runner.js:276:5)
    at processImmediate [as _immediateCallback] (timers.js:345:15)
```

2.0.1 works fine. As the [wiki page](https://github.com/mochajs/mocha/wiki/Using-mocha-programmatically) shows use of `mocha.run()`, I think it's a regression in 2.1.0. A workaround is [to pass an empty callback](https://github.com/vibe-project/vibe-javascript-client/commit/058f9d5020306cf2800e9f2862b4a773637483c5).

It looks like that https://github.com/mochajs/mocha/commit/30582e646d579ce67ef4229922651da498e79a80 is the cause. Before executing `fn(failures);`, it should check if `fn` is available.

### PR DESCRIPTION
**Pull request by dasilvacontin**
See issue #1496.

### CODE REVIEW

**Comment by a8m** [comment]
I just came across with this issue yesterday, and planning to fix it.
but this pr looks fair enough, so I'm +1
also, it will be great if you add tests for this too.

**Comment by dasilvacontin** [comment]
@a8m, true, thanks for reminding me of the tests.

**Comment by dasilvacontin** [comment]
Added some tests. Any other test you'd add?

File: test/mocha.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+        it('should execute the callback with the number of failures '+
+            'as parameter', function(done) {
+            var mocha = new Mocha(blankOpts)
+            mocha.suite.addTest(new Test('failing test', function(){
+                throw new Error('such fail')
+            }))
+            mocha.run(function(failures) {
+                failures.should.equal(1)
+                done()
```
Comment (travisjeffery) [inline test/mocha.js]: use semicolons

File: test/mocha.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+var Mocha = require('../')
+  , Test = Mocha.Test
```
Comment (travisjeffery) [inline test/mocha.js]: use multiple var statements

File: test/mocha.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+describe('Mocha', function(){
+    var blankOpts = { reporter: function(){}} // no output
```
Comment (travisjeffery) [inline test/mocha.js]: 2 space indentation

Comment (travisjeffery) [inline test/mocha.js]: 2 space indentation

**Comment by travisjeffery** [comment]
made some comments there to fix and also squash your commits, thanks!

**Comment by dasilvacontin** [comment]
We really need JSHint or similar. I'll make the changes asap, thanks for pointing out, @travisjeffery!

Commits were squash, and will squash after new changes.

**Comment by dasilvacontin** [comment]
Changes done. /cc @travisjeffery

**Comment by travisjeffery** [comment]
@dasilvacontin i just rechecked the current tests we have, and pretty much all the tests you added are dups of the tests here: https://github.com/mochajs/mocha/blob/master/test/runnable.js#L92

**Comment by dasilvacontin** [comment]
@travisjeffery, those tests are for `Runnable`, while mines are for `Mocha`.

**Comment by dasilvacontin** [comment]
Thanks @travisjeffery!

**Pull request merged**